src/crystal/hasher.cr

```diff
@@ -45,12 +94,53 @@ struct Crystal::Hasher
   end
 
   def reference(value)
-    @result = @result * 31 + value.object_id.to_u64
-    self
+    permute(value.object_id.to_u64)
   end
 
   def string(value)
-    value.to_slice.hash(self)
+    bytes(value.to_slice)
+  end
+
+  private def read_u24(ptr, rest)
+    ptr[0].to_u64 | (ptr[rest/2].to_u64 << 8) | (ptr[rest - 1].to_u64 << 16)
+  end
+
+  private def read_u32(ptr)
+    # force correct unaligned read
+    t4 = uninitialized UInt32
+    pointerof(t4).as(UInt8*).copy_from(ptr, 4)
+    t4.to_u64
+  end
+
+  private def read_u64(ptr)
+    # force correct unaligned read
+    t8 = uninitialized UInt64
+    pointerof(t8).as(UInt8*).copy_from(ptr, 8)
+    t8
+  end
+
+  def bytes(value)
```

Comment:
Since `Hasher#bytes` is being used in other places (as the method is public), can we get some documentation comments on _when_ to use `#bytes` instead of `hasher.hash`?

Thank you.

Reply:
May be it should be returned back to `def bytes(value : Bytes)` ?
Or even changed to `def bytes(ptr : Pointer(UInt8), size : UInt64)` ?
I'd prefer last variant.

Reply:
Bytes is just a pointer and size so there's no point to the last variant.

Reply:
Ok, I did my best: added comments to `Hasher#bytes(value : Bytes)` and to `Slice(UInt8).hash(hasher)`.